### PR TITLE
Feature/storage power actor

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
### Description of the Change
HAMT-based Power Table.

### CODE REVIEW

File: core/power/impl/power_table_hamt.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  return result;
+}
+
+fc::outcome::result<void> PowerTableHamt::setMinerPower(const Address &address,
+                                                        Power power_amount) {
+  if (power_amount < 0) return PowerTableError::NEGATIVE_POWER;
+  return power_table_.setCbor(encodeToByteString(address), power_amount);
+}
+
+/** @copydoc PowerTable::getMinerPower() */
```
Comment (artyom-yurin) [inline core/power/impl/power_table_hamt.cpp]:
```suggestion
/** @copydoc PowerTable::removeMiner() */
```
Also others functions

File: core/vm/actor/builtin/storage_power/storage_power_actor_state.cpp [hunk 2]
```diff
@@ -3,26 +3,35 @@
...
+  StoragePowerActorState::StoragePowerActorState(
+      std::shared_ptr<Indices> indices,
+      std::shared_ptr<crypto::randomness::RandomnessProvider>
+          randomness_provider)
+      : indices_(std::move(indices)),
+        randomness_provider_(std::move(randomness_provider)),
+        total_network_power_(0),
+        power_table_(std::make_unique<power::PowerTableImpl>()),
+        claimed_power_(std::make_unique<power::PowerTableImpl>()),
+        nominal_power_(std::make_unique<power::PowerTableImpl>()),
```
Comment (artyom-yurin) [inline core/vm/actor/builtin/storage_power/storage_power_actor_state.cpp]:
```suggestion
        nominal_power_(std::make_unique<power::PowerTableHamt>()),
```
Also for two lines above

File: core/vm/actor/builtin/storage_power/storage_power_actor_state.cpp [hunk 3]
```diff
@@ -3,26 +3,35 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#include "vm/actor/builtin/storage_power/storage_power_actor.hpp"
+#include "vm/actor/builtin/storage_power/storage_power_actor_state.hpp"
 
 #include "power/impl/power_table_impl.hpp"
```
Comment (artyom-yurin) [inline core/vm/actor/builtin/storage_power/storage_power_actor_state.cpp]:
```suggestion
#include "power/impl/power_table_hamt.hpp"
```

File: core/vm/actor/builtin/storage_power/CMakeLists.txt [hunk 4]
```diff
@@ -4,7 +4,7 @@
 #
 
 add_library(storage_power_actor
-    storage_power_actor.cpp
+    storage_power_actor_state.cpp
     )
 target_link_libraries(storage_power_actor
```
Comment (artyom-yurin) [inline core/vm/actor/builtin/storage_power/CMakeLists.txt]: Change `power_table` to `power_table_hamt`

File: core/power/CMakeLists.txt [hunk 5]
```diff
@@ -2,10 +2,20 @@
 # SPDX-License-Identifier: Apache-2.0
 
 add_library(power_table
```
Comment (artyom-yurin) [inline core/power/CMakeLists.txt]: Should be removed

File: core/power/impl/power_table_hamt.cpp (hunk 1 above)
Comment (masterjedy) [reply core/power/impl/power_table_hamt.cpp]: we need docs in headers, not in sources

Comment (masterjedy) [reply core/power/impl/power_table_hamt.cpp]: move descriptions to headers

File: core/power/impl/power_table_impl.cpp [hunk 6]
```diff
@@ -36,11 +36,11 @@ namespace fc::power {
     return outcome::success();
   }
 
-  size_t PowerTableImpl::getSize() const {
+  fc::outcome::result<size_t> PowerTableImpl::getSize() const {
```
Comment (masterjedy) [inline core/power/impl/power_table_impl.cpp]: I don't think we should wrap it with outcome::result, it makes no sense.

Comment (masterjedy) [reply core/power/impl/power_table_impl.cpp]: I see, it's due to hamt.

File: core/primitives/big_int.hpp [hunk 7]
```diff
@@ -33,12 +35,24 @@ namespace fc::primitives {
...
+
+    inline UBigInt operator-=(const UBigInt &other) {
+      static_cast<cpp_int *>(this)->operator-=(static_cast<cpp_int>(other));
+      return *this;
     }
 
-    inline UBigInt operator-(const UBigInt &other) const {
-      return static_cast<cpp_int>(*this) - static_cast<cpp_int>(other);
+    inline UBigInt operator*=(const UBigInt &other) {
+      static_cast<cpp_int *>(this)->operator*=(static_cast<cpp_int>(other));
```
Comment (masterjedy) [inline core/primitives/big_int.hpp]: why not
```suggestion
      (*static_cast<cpp_int *>(this)) *= static_cast<cpp_int>(other);
```

Comment (masterjedy) [reply core/primitives/big_int.hpp]: same for other places nearby

File: core/vm/actor/builtin/storage_power/storage_power_actor_state.cpp (hunk 2 above)
Comment (Alexey-N-Chernyshov) [reply core/vm/actor/builtin/storage_power/storage_power_actor_state.cpp]: Not yet, the state should be reworked (up-to date specs-actor, CBOR-serialisation, etc.)

File: core/power/impl/power_table_hamt.cpp (hunk 1 above)
Comment (Alexey-N-Chernyshov) [reply core/power/impl/power_table_hamt.cpp]: fixed

File: core/primitives/big_int.hpp (hunk 7 above)
Comment (Alexey-N-Chernyshov) [reply core/primitives/big_int.hpp]: done

File: core/power/CMakeLists.txt (hunk 5 above)
Comment (Alexey-N-Chernyshov) [reply core/power/CMakeLists.txt]: The state is not ready yet, I will replace in the next PR with proper reworking of state (use HAMT, AMT, structure up-to specs-actor, CBOR-ing).

**Pull request merged**